### PR TITLE
Fixes on AMQP received message flow.

### LINKED ISSUE #25: cloud_receive_message is not been called when return is false
If the callback set in function `l_io_set_read_handler` returns false it will destroy the read_handler (aka. on_receive:amqp.c)

### PR DESCRIPTION
**Pull request by vitbaq**
Closes #25

### CODE REVIEW

File: src/amqp.c [hunk 1]
```diff
@@ -140,15 +140,17 @@ static bool on_receive(struct l_io *io, void *user_data)
 	body = amqp_bytes_to_new_string(envelope.message.body);
 
 	err = amqp_ctx.read_cb(exchange, routing_key, body, user_data);
+	if (!err)
+		hal_log_dbg("Invalid message to the assigned callback");
+		/* FIXME: Put msg on the queue again */
```
Comment (ramonhpr) [inline src/amqp.c]: Put comment before the command in the commit message

File: src/msg.c [hunk 2]
```diff
@@ -1431,7 +1431,7 @@ static void append_queue(void *data, void *user_data)
 	knot_msg_data *msg = data;
 
 	l_queue_push_tail(session->update_list,
-			  l_memdup(msg, sizeof(msg)));
+			  l_memdup(msg, sizeof(*msg)));
```
Comment (ramonhpr) [inline src/msg.c]: Explain what is wrong on the commit

File: src/amqp.c [hunk 3]
```diff
@@ -331,7 +331,7 @@ int amqp_set_read_cb(amqp_read_cb_t on_read, void *user_data)
 	amqp_ctx.amqp_io = l_io_new(amqp_get_sockfd(amqp_ctx.conn));
 
 	err = l_io_set_read_handler(amqp_ctx.amqp_io, on_receive,
-				    &amqp_ctx.conn, user_data);
+				    user_data, NULL);
```
Comment (ramonhpr) [inline src/amqp.c]: Explain the two parameters in the commit message.

File: src/cloud.c [hunk 4]
```diff
@@ -69,32 +69,30 @@ static bool cloud_receive_message(const char *exchange,
...
 	const char *id;
+	bool result = true;
 
 	if (strcmp(AMQP_EXCHANGE_FOG, exchange) != 0)
 		return true;
 
 	jso = json_tokener_parse(body);
 	if (!jso)
-		return false;
+		goto error;
```
Comment (ramonhpr) [inline src/cloud.c]: goto reject_msg

File: src/cloud.c [hunk 5]
```diff
@@ -105,6 +103,13 @@ static bool cloud_receive_message(const char *exchange,
 	}
 
 	json_object_put(jso);
+	return result;
+
+error:
```
Comment (ramonhpr) [inline src/cloud.c]: reject_msg

File: src/cloud.c [hunk 6]
```diff
@@ -69,32 +69,30 @@ static bool cloud_receive_message(const char *exchange,
...
 		return true;
 
 	jso = json_tokener_parse(body);
 	if (!jso)
-		return false;
+		goto error;
 
 	if (!json_object_object_get_ex(jso, "id", &id_json))
-		return false;
+		goto error;
```
Comment (ramonhpr) [inline src/cloud.c]: Make a function `get_id_from_json_obj`

File: src/cloud.c [hunk 7]
```diff
@@ -69,32 +69,30 @@ static bool cloud_receive_message(const char *exchange,
...
 	if (cloud_cbs.update_cb != NULL &&
 	    strcmp(AMQP_EVENT_DATA_UPDATE, routing_key) == 0) {
 		list = parser_update_to_list(jso);
-		if (list == NULL) {
-			hal_log_error("Error on parse json object");
-			json_object_put(jso);
-			return false;
-		}
+		if (list == NULL)
+			goto error;
```
Comment (ramonhpr) [inline src/cloud.c]: goto reject_msg

File: src/amqp.c [hunk 8]
```diff
@@ -140,15 +140,17 @@ static bool on_receive(struct l_io *io, void *user_data)
...
 
-	/* FIXME: Put msg on the queue again */
 	hal_log_dbg("Destroy received envelope");
 	amqp_destroy_envelope(&envelope);
 	l_free(exchange);
 	l_free(routing_key);
 	l_free(body);
 
-	return err;
+	return true;
```
Comment (ramonhpr) [inline src/amqp.c]: Explain why we should return true here

File: src/cloud.c (hunk 4 above)
Comment (vitbaq) [reply src/cloud.c]: Changed to "done" in the new context.

File: src/cloud.c (hunk 7 above)
Comment (vitbaq) [reply src/cloud.c]: Changed to "done" in the new context.

File: src/cloud.c (hunk 5 above)
Comment (vitbaq) [reply src/cloud.c]: Changed to "done" in the new context.

File: src/amqp.c [hunk 9]
```diff
@@ -140,15 +145,17 @@ static bool on_receive(struct l_io *io, void *user_data)
 	body = amqp_bytes_to_new_string(envelope.message.body);
 
 	err = amqp_ctx.read_cb(exchange, routing_key, body, user_data);
```
Comment (alla3) [inline src/amqp.c]: Change the retuning variable name. I suggest **success** instead of **err**.

File: src/cloud.c [hunk 10]
```diff
@@ -61,40 +61,62 @@ struct cloud_callbacks {
...
+ */
 static bool cloud_receive_message(const char *exchange,
 				  const char *routing_key,
 				  const char *body, void *user_data)
 {
 	json_object *jso;
-	json_object *id_json;
 	struct l_queue *list;
 	const char *id;
+	bool discard = true;
```
Comment (alla3) [inline src/cloud.c]: Change the name of this variable, I suggest **consumed** instead of **discard**.

File: src/amqp.c [hunk 11]
```diff
@@ -139,16 +144,18 @@ static bool on_receive(struct l_io *io, void *user_data)
 	routing_key = amqp_bytes_to_new_string(envelope.routing_key);
 	body = amqp_bytes_to_new_string(envelope.message.body);
 
-	err = amqp_ctx.read_cb(exchange, routing_key, body, user_data);
+	success = amqp_ctx.read_cb(exchange, routing_key, body, user_data);
+	if (!success)
+		/* FIXME: Put msg on the queue again */
```
Comment (alla3) [inline src/amqp.c]: I think it is more like a TODO than a FIXME.
Also, I would make changes on the comment.

~Put~ Add **the** message...

File: src/amqp.c (hunk 3 above)
Comment (ramonhpr) [reply src/amqp.c]: Nice

File: src/amqp.c [hunk 12]
```diff
@@ -102,13 +102,18 @@ static char *amqp_bytes_to_new_string(amqp_bytes_t data)
 	return str;
 }
 
+/**
+ * Callback function to consume message envelope from AMQP queue.
+ *
+ * Returns true on success or false if the read callback is not set.
```
Comment (ramonhpr) [inline src/amqp.c]: This comment doesn't seem to be related to this callback. Shouldn't `on_receive` returns true always so we avoid that `l_io_set_read_handler` stop calling it?

File: src/amqp.c [hunk 13]
```diff
@@ -102,13 +102,18 @@ static char *amqp_bytes_to_new_string(amqp_bytes_t data)
 	return str;
 }
 
+/**
```
Comment (ramonhpr) [inline src/amqp.c]: Related to commit message 69b2b1a7c8f056a214d21424e8298ad8d9b4cb06:
Remove `user`. There is no user interaction on this flow. Use publisher/consumer or client.

File: src/amqp.c (hunk 12 above)
Comment (vitbaq) [reply src/amqp.c]: It is related to line 135.

**Pull request merged**